eeps/eep-0057.md

```diff
@@ -0,0 +1,353 @@
+    Author: José Valim <jose(dot)valim(at)gmail(dot)com>
+    Status: Draft
+    Type: Standards Track
+    Created: 02-Jun-2021
+    Post-History:
+****
+EEP 57: Module attributes for documentation
+----
+
+Abstract
+========
+
+This EEP proposes a structured documentation API for Erlang where the
+documentation is handled as part of the language parser and included
+directly in the compiled .beam files, as a replacement for EDoc.
+Python, Elixir, and Clojure are examples of languages that follow this
+approach of treating documentation as data rather than code comments.
+
+Rationale
+=========
+
+The main limitation in EDoc today is that the documentation is kept as
+code comments.  This requires an explicit tool to parse said code
+comments, which complicates access to the docs by IDEs, from the shell,
+etc.  There have been recent improvements in this area by making EDoc
+compile to EEP 48 but it still requires an explicit step.
+
+Furthermore, the “code comments” approach is more complex implementation
+wise, as it requires parsing the source code alongside code comments,
+parsing the code comments, and so on.  Personally, I also enjoy the
+explicit distinction between documentation and code comments: they have
+different requirements and different audiences.
+
+This EEP proposes the addition of two module attributes to Erlang: `-doc`
+and `-moduledoc`.  It also includes an extra section on additional features
+that can make aid writing documentation, but they are optional and some
+of them should likely be explored more in depth in other EEPs.
+Still, they are included to provide a long term view of the features and
+challenges related to structured documentation.
+          
+As with EEP 48, this proposal pertains exclusively to API references and
+their documentation.  It doesn't cover guides, tutorials, and other
+documentation formats.
+
+New module attributes
+=====================
+
+This EEP proposes two new attributes: `-doc` and `-moduledoc`.  They could
+be used as follows:
+
+    -module(base64).
+    -moduledoc "
+    Convenience functions for encoding and decoding from base64.
+    ".
+
+    -doc "
+    Encodes the given binary to base64.
+    ".
+    -spec encode(binary()) -> binary().
+    encode(Binary) ->
+      % ....
+
+    -doc "
+    Decodes the given binary from base64.
+    ".
+    -spec decode(binary()) -> {ok, binary()} | error.
+    encode(Binary) ->
+      % ....
+
+The new `-moduledoc` attribute can be listed anywhere and it will contain
+the documentation for the given module.  The `-doc` attribute must be
+listed once before each function and it will contain the documentation
+for the following function.
```

Comment:
Any ordering preferences compared to type specs?

Reply:
@ferd please expand? I am not sure I follow the question. :)

Reply:
People have a tendency to do something like:

```
%% @doc This function turns the programmer into a much more productive one
-spec levelup(programmer()) -> programmer().
levelup(Programmer) -> Programmer*10.
```

The new literal translation would be:

```
-doc "This function turns the programmer into a much more productive one".
-spec levelup(programmer()) -> programmer().
levelup(Programmer) -> Programmer*10.
```

This will work with the spec above if we read "listed _once before_ each function", meaning there can be one `-doc` attribute before any function and it will get assigned to the one that follows.

However the spec could also force the following if we read it to mean "the doc must be listed once _right before_ each function":

```
-spec levelup(programmer()) -> programmer().
-doc "This function turns the programmer into a much more productive one".
levelup(Programmer) -> Programmer*10.
```

Reply:
Ah, I see. We don't look at other attributes, so it can be anywhere before a function. I will clarify.